File: smart_trip_ai_mobile/lib/services/geocoding_service.dart
import 'dart:convert';
import 'dart:math';
import 'package:http/http.dart' as http;
import '../domain/models/destination.dart';

// Cache for live geocoded results
final Map<String, List<Destination>> _geocodeCache = {};

/// Live geocoding search for ANY location, street, town, hotel or landmark in Sri Lanka using OpenStreetMap Nominatim API.
Future<List<Destination>> searchSriLankaLocations(String query,
    {double currentLat = 6.9271, double currentLng = 79.8612}) async {
  final trimmed = query.trim();
  if (trimmed.isEmpty || trimmed.length < 2) return [];

  final cacheKey = trimmed.toLowerCase();
  if (_geocodeCache.containsKey(cacheKey)) {
    return _geocodeCache[cacheKey]!;
  }

  try {
    final queryUrl = Uri.parse(
        'https://nominatim.openstreetmap.org/search?format=json&q=${Uri.encodeComponent('$trimmed, Sri Lanka')}&countrycodes=lk&limit=8');
    final response = await http.get(queryUrl, headers: {
      'Accept-Language': 'en-US,en;q=0.9',
      'User-Agent': 'SmartTripAI Telematics App (Sri Lanka)',
    });

    if (response.statusCode == 200) {
      final List<dynamic> data = json.decode(response.body);
      final List<Destination> results = [];

      for (int i = 0; i < data.length; i++) {
        final item = data[i];
        final lat = double.parse(item['lat'].toString());
        final lon = double.parse(item['lon'].toString());
        
        final nameParts = item['display_name'].toString().split(',');
        final mainName = nameParts.isNotEmpty ? nameParts[0].trim() : item['display_name'];
        final cityName = nameParts.length > 1 
          ? nameParts[1].trim() 
          : (nameParts.length > 2 ? nameParts[2].trim() : 'Sri Lanka');

        // Calculate Haversine distance from current position
        final distKm = _calculateDistanceKm(currentLat, currentLng, lat, lon);

        results.add(Destination(
          id: 'geo-$i-${DateTime.now().millisecondsSinceEpoch}',
          name: mainName,
          city: cityName,
          latitude: lat,
          longitude: lon,
          distanceKmEst: double.parse(distKm.toStringAsFixed(1)),
        ));
      }

      _geocodeCache[cacheKey] = results;
      return results;
    }
  } catch (error) {
    print('Live geocoding search error: $error');
  }

  return [];
}

double _calculateDistanceKm(double lat1, double lon1, double lat2, double lon2) {
  const r = 6371.0;
  final dLat = ((lat2 - lat1) * pi) / 180;
  final dLon = ((lon2 - lon1) * pi) / 180;
  final a = sin(dLat / 2) * sin(dLat / 2) +
      cos((lat1 * pi) / 180) *
          cos((lat2 * pi) / 180) *
          sin(dLon / 2) *
          sin(dLon / 2);
  final c = 2 * atan2(sqrt(a), sqrt(1 - a));
  return r * c;
}

File: src/services/geocodingService.ts
import type { Destination } from '../presentation/components/DestinationSearchBar';

// Cache for live geocoded results
const geocodeCache = new Map<string, Destination[]>();

/**
 * Live geocoding search for ANY location, street, town, hotel or landmark in Sri Lanka using OpenStreetMap Nominatim API.
 */
export async function searchSriLankaLocations(query: string, currentLat: number = 6.9271, currentLng: number = 79.8612): Promise<Destination[]> {
  const trimmed = query.trim();
  if (!trimmed || trimmed.length < 2) return [];

  const cacheKey = trimmed.toLowerCase();
  if (geocodeCache.has(cacheKey)) {
    return geocodeCache.get(cacheKey)!;
  }

  try {
    const url = `https://nominatim.openstreetmap.org/search?format=json&q=${encodeURIComponent(trimmed + ', Sri Lanka')}&countrycodes=lk&limit=8`;
    const response = await fetch(url, {
      headers: {
        'Accept-Language': 'en-US,en;q=0.9',
        'User-Agent': 'SmartTripAI Telematics App (Sri Lanka)',
      },
    });

    if (response.ok) {
      const data = await response.json();
      const results: Destination[] = data.map((item: any, idx: number) => {
        const lat = parseFloat(item.lat);
        const lon = parseFloat(item.lon);
        const nameParts = item.display_name.split(',');
        const mainName = nameParts[0]?.trim() || item.display_name;
        const cityName = nameParts[1]?.trim() || nameParts[2]?.trim() || 'Sri Lanka';

        // Calculate Haversine distance from current position
        const distKm = calculateDistanceKm(currentLat, currentLng, lat, lon);

        return {
          id: `geo-${idx}-${Date.now()}`,
          name: mainName,
          city: cityName,
          latitude: lat,
          longitude: lon,
          distanceKmEst: parseFloat(distKm.toFixed(1)),
        };
      });

      geocodeCache.set(cacheKey, results);
      return results;
    }
  } catch (error) {
    console.warn('Live geocoding search error:', error);
  }

  return [];
}

function calculateDistanceKm(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLon / 2) *
      Math.sin(dLon / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}
